### ISSUE TITLE
Installing VMOD on alpine:3.7 fails on make

### ISSUE THREAD

**Issue by LMalikov** [carlosabalde/libvmod-cfg#30]
Varnish version is 5.2.1-r0, cfg branch is 4.1.
When trying to install the VMOD during make I get:
`In file included from remote.c:11:0:
remote.c: In function 'check_remote':
helpers.h:40:35: error: format '%s' expects argument of type 'char *', but argument 8 has type 'int' [-Werror=format=]
             VSLb((ctx)->vsl, slt, "[CFG][%s:%d] " fmt, __func__, __LINE__, __VA_ARGS__); \
                                   ^
remote.c:186:25: note: in expansion of macro 'LOG'
                         LOG(ctx, LOG_ERR,
                         ^~~
helpers.h:42:25: error: format '%s' expects argument of type 'char *', but argument 8 has type 'int' [-Werror=format=]
             VSL(slt, 0, "[CFG][%s:%d] " fmt, __func__, __LINE__, __VA_ARGS__); \
                         ^
remote.c:186:25: note: in expansion of macro 'LOG'
                         LOG(ctx, LOG_ERR,
                         ^~~
cc1: all warnings being treated as errors
make[2]: *** [Makefile:528: remote.lo] Error 1
make[2]: Leaving directory '/tmp/libvmod-cfg/src'
make[1]: *** [Makefile:536: all-recursive] Error 1
make[1]: Leaving directory '/tmp/libvmod-cfg'
make: *** [Makefile:404: all] Error 2`

What could I be missing?

**Comment by carlosabalde** [comment]
Branch `4.1` should be only used with Varnish Cache 4.1.x. Varnish 5.2.x is EOL and it's not supported by the VMOD. You can try `5.2-archived` branch. It should work but that's an unmaintained branch.

**Comment by LMalikov** [comment]
I downgraded to alpine 3.6 and Varnish 4.1.9, still getting the same error on make :(

**Comment by carlosabalde** [comment]
Please, ensure you are trying to build it using a clean download (e.g. `git clean -f -x -d`), and then run `autogen.sh`, `configure` and `make`.

**Comment by LMalikov** [comment]
Still no luck, same error. I'm trying to include the VMOD in a Docker container, here's the Dockerfile:
`FROM alpine:3.6

RUN apk update
ARG VARNISH_VERSION=4.1.9-r0

RUN \
  apk update && \
  apk upgrade && \
  apk add --update ca-certificates libressl && \
  update-ca-certificates && \
  apk add --no-cache varnish=${VARNISH_VERSION} varnish-dev=${VARNISH_VERSION} \
    varnish-libs=${VARNISH_VERSION}

RUN \
  apk add --no-cache autoconf automake libtool make py-docutils python git 

# Add cfg vmod and dependencies
RUN \
  apk add --no-cache curl-dev luajit-dev && \
  cd /tmp && git clone --branch 4.1 https://github.com/carlosabalde/libvmod-cfg.git && \
  cd libvmod-cfg && git clean -f -x -d && ./autogen.sh && ./configure && make && make install

**Comment by carlosabalde** [comment]
Thanks. That's helpful. The issue is related with `strerror_r` and Alpine Linux (POSIX vs. GNU version).  Not sure about the right solution yet.

**Comment by carlosabalde** [comment]
I've pushed a few changes to all supported branches that should fix the [Linux Alpine compilation issue](https://stackoverflow.com/questions/41953104/strerror-r-is-incorrectly-declared-on-alpine-linux) (basically, unconditionally stop using `strerror_r` 😢). I've tried on my own and compilation now completes successfully (although you need some extra packages: `apk add --no-cache linux-headers jemalloc-dev vim`). However, I still see issues if you try to run tests. Not sure if you will experience the same problems when using the VMOD. I'll keep this open if you want to share your experience.

**Comment by LMalikov** [comment]
Thanks for taking time to look into it, Carlos :)
The VMOD gets installed now, but import in VCL fails with the following:
```Error:
Message from VCC-compiler:
Could not load VMOD cfg
	File name: libvmod_cfg.so
	dlerror: Error relocating /usr/lib/varnish/vmods/libvmod_cfg.so: mallctl: symbol not found
('/etc/varnish/default.vcl' Line 9 Pos 8)
import cfg;
-------###-```

**Comment by carlosabalde** [comment]
That's exactly the problem I expected. You will get the same one if you try to run tests with `make check`.

I guess that's a `musl` vs. `glibc` issue, but it'd require more time and effort to really understand what's going on and to find the right way to properly support Alpine Linux. I'll keep this open, but personally I simply avoid using an Alpine based container for this.

**Comment by carlosabalde** [comment]
We've been looking at this today:

- We've identified a minor permissions-related issue when running tests in Linux Alpine. It has been fixed with 911f38ddfb4f6143a1de3b079b3ad410a871b80d.

- We've also found the reason for `mallctl: symbol not found` errors. It's related with a `musl` vs. `glibc` thing, in particular the issue is related with a `jemalloc` optimization included in the VMOD. Good news is you can disable it using `./configure --disable-flush-jemalloc-tcache`, compile the VMOD again, and it should load as expected (and that way you should be able to run tests too).

Again, personally I'd avoid using an Alpine based container for Varnish Cache, but at lest now you should be able to compile and load the VMOD.

Huge thank you for @davidfb! He's the one who found these issues. I'm here just for the glory 🎉
